ha_lineairdb.cc

```diff
@@ -751,107 +747,85 @@ int ha_lineairdb::is_primary_key_type_int() {
+        current_key_ += ldbField.get_lineairdb_field();
       }
-    } else if (is_primary_key_exists()) {  // KEY EXISTS
-      assert(max_supported_key_parts() ==
-             1);  // now we assume that there is no composite index
+    }
+    else if (is_primary_key_exists()) {  // KEY EXISTS
+      // now we assume that there is no composite index
+      assert(max_supported_key_parts() == 1);
       my_bitmap_map* org_bitmap = tmp_use_all_columns(table, table->read_set);
       for (Field** field = table->field; *field; field++) {
         auto* f = *field;
         if (f->m_indexed) {  // it is the key column
-
           String b;
           (*field)->val_str(&b, &b);
-          current_key_.append(b);
+          ldbField.set_lineairdb_field(b.c_ptr(), b.length());
+          current_key_ += ldbField.get_lineairdb_field();
           break;
         }
       }
       tmp_restore_column_map(table->read_set, org_bitmap);
-    } else {  // THERE IS NO KEY COLUMN
+    } 
+    else {  // THERE IS NO KEY COLUMN
       const auto cstr       = table->s->table_name;
       const auto table_name = std::string(cstr.str, cstr.length);
-      auto inserted_count   = auto_generated_keys_[table_name]++;
-      current_key_.append(std::to_string(inserted_count));
+      auto inserted_count = auto_generated_keys_[table_name]++;
+      std::string&& s = std::to_string(inserted_count);
+      ldbField.set_lineairdb_field(s.c_str(), s.size());
+      current_key_ += ldbField.get_lineairdb_field();
     }
   }
 }
 
-std::string ha_lineairdb::get_current_key() {
-  return std::string(current_key_.ptr(), current_key_.length());
-}
-
-inline bool bit_is_up(uchar& flag, size_t idx){
-  return (flag >> idx) & 1;
-}
+std::string ha_lineairdb::get_current_key() { return current_key_; }
 
 /**
  * @brief Format and set the requested row into `write_buffer_`.
- * A MySQL Row is serialized to comma-separated double-quoted values,
- * as the followings:
- * {"alice","bob","carol","3"}
- *   col1    col2  col3   col4
  */
 void ha_lineairdb::set_write_buffer(uchar* buf) {
-  write_buffer_.length(0);
+  ldbField.set_lineairdb_field(buf, table->s->null_bytes);
+  write_buffer_ = ldbField.get_lineairdb_field();
 
   char attribute_buffer[1024];
   String attribute(attribute_buffer, sizeof(attribute_buffer), &my_charset_bin);
 
-  translator.reset();
-
   my_bitmap_map* org_bitmap = tmp_use_all_columns(table, table->read_set);
   for (Field** field = table->field; *field; field++) {
-    const char* p;
-    const char* end;
-
     (*field)->val_str(&attribute, &attribute);
-    p   = attribute.ptr();
-    end = p + attribute.length();
-
-    write_buffer_.append('"');
-    for (; p < end; p++) write_buffer_.append(*p);
-    write_buffer_.append('"');
-    write_buffer_.append(',');
-    if ((*field)->is_nullable()) translator.check_flag_length();
+    ldbField.set_lineairdb_field(attribute.c_ptr(), attribute.length());
```

Comment:
The function `set_lineairdb_field` receives two-parameter patterns as followings… is this correct?
1. a uchar of mysql row format (`uchar buf`), and a bitmap of null flags.
2. a pointer of each field (`attribute`) and the length.

If so, it's a bit confusing. I am confused by this function appearing twice (L816 and L825) in a few lines but having different usages.
The current function name gives the image of the latter (2.) usage but the former is not.


It is better to separate the function names, even if the internal procedures are the same.

Reply:
I agree to separate function names. Currently the function is doing two different things.
I created new functions to store null flags in LineairDB.